Add action to initialize time-dependent DG domain

nilsdeppe opened the pull request:
## Proposed changes

- Add action to initialize the DG domain for the evolution code. This needs to be separate from the elliptic initialization action since the elliptic code doesn't need time-dependent domains. 

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add action to initialize time-dependent DG domain

kidder commented on src/Evolution/Initialization/DgDomain.hpp:
```diff
@@ -0,0 +1,176 @@
+
+    using compute_tags = db::AddComputeTags<
+        ::domain::Tags::LogicalCoordinates<Dim>,
+
```
add comments instead of blank lines for groupings

kidder commented on tests/Unit/Evolution/Initialization/Test_DgDomain.cpp:
```diff
@@ -0,0 +1,499 @@
+    }
+  };
+
+  const auto check_domain_tags_time_independent =  //
```
why the `//` ?

kidder reviewed (changes requested):
squash

moxcodes commented on src/Evolution/Initialization/DgDomain.hpp:
```diff
@@ -0,0 +1,176 @@
+ *
+ * ConstGlobalCache:
+ * - Uses:
+ *   - `Tags::Domain<Dim, Frame::Inertial>`
```
In case it's worth addressing, none of these tags link in doxygen

moxcodes commented on src/Evolution/Initialization/DgDomain.hpp:
```diff
@@ -0,0 +1,176 @@
+ *   - `Tags::Domain<Dim, Frame::Inertial>`
+ * DataBox:
+ * - Uses:
+ *   - `Tags::InitialExtents<Dim>`
```
You also use `InitialFunctionsOfTime<Dim>`.
edit: I think the `Adds:` list similarly needs updating

moxcodes commented on src/Evolution/Initialization/DgDomain.hpp:
```diff
@@ -0,0 +1,176 @@
+        std::string, std::unique_ptr<::domain::FunctionsOfTime::FunctionOfTime>>
+        functions_of_time{};
+    for (const auto& name_and_function : initial_functions_of_time) {
+      functions_of_time[name_and_function.first] =
```
[optional] could be `insert` to be more explicit

moxcodes commented on tests/Unit/Evolution/Initialization/Test_DgDomain.cpp:
```diff
@@ -0,0 +1,500 @@
+
+  PUPable_reg(SINGLE_ARG(TimeIndependentMap<Dim, Frame::Inertial>));
+  PUPable_reg(SINGLE_ARG(TimeIndependentMap<Dim, Frame::Grid>));
+  PUPable_reg(SINGLE_ARG(TimeDependentMap<Dim>));
```
does this one need to be `SINGLE_ARG`?

moxcodes commented on tests/Unit/Evolution/Initialization/Test_DgDomain.cpp:
```diff
@@ -0,0 +1,500 @@
+  std::unordered_map<std::string,
+                     std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
+      functions_of_time{};
+  functions_of_time[functions_of_time_names[0]] =
```
could probably be a loop over the thee indexed assignments.

moxcodes commented on tests/Unit/Evolution/Initialization/Test_DgDomain.cpp:
```diff
@@ -0,0 +1,500 @@
+                     create_affine_map<Dim, Frame::Inertial>()),
+                 0,
+                 {}};
+  Domain<Dim> domain{std::move(blocks)};
```
can the `Block<Dim>` constructor just be put into an initializer list inside this constructor (with some indeterminate number of `{`)?

nilsdeppe replied:
No... C++ is dumb and doesn't do move semantics correctly with initializer lists and so it requires that `blocks` be copyable, which it isn't since it holds `unique_ptr`s

moxcodes replied:
Oh, right... Will just told us about that recently. Such a frustrating clash in the standard lib :/

nilsdeppe committed:
Add dox for evolution::dg::(Actions|Initialization)

nilsdeppe commented:
Did a few things:
- Rebased on develop
- Added a commit that documents `evolution::dg`, `evolution::dg::Actions`, and `evolution::dg::Initialization`
- Moved the action to `evolution::dg::Initialization` and renamed it to `Domain`
- Documented the tags (hopefully all and correctly)
- Switched to `insert` where possible

moxcodes reviewed (commented):
Looks good, go ahead and squash

nilsdeppe commented:
rebased on develop again and squashed
